demos/dotnet/dotnet/ProjectPropertiesHelpers.cs

```diff
@@ -0,0 +1,335 @@
+        {
+            foreach (var file in files.Where(file => !list.Contains(file)))
+            {
+                list.Add(file);
+            }
+        }
+
+        private static void AddToListWithoutDuplicates(ICollection<string> list, string file)
+        {
+            if (!list.Contains(file))
+            {
+                list.Add(file);
+            }
+        }
+
+        private static void AddToListWithoutDuplicates(ICollection<string> list, IEnumerable<string> files)
+        {
+            foreach (var file in files.Where(file => !list.Contains(file)))
+            {
+                list.Add(file);
+            }
+        }
+
+        private static void LogProperties(this Log log, ProjectProperties project, string heading, bool buildDll)
+        {
+            if (!log.IsEnabled) return;
+
+            log.WriteLine(heading);
+            log.WriteLine("ProjectDirectory     {0}", project.ProjectDirectory);
+            log.WriteLine("PackagesDirectory    {0}", project.PackagesDirectory);
+            log.WriteLine("OutputDirectory      {0}", project.OutputDirectory);
+            log.WriteLine("ToolsDirectory       {0}", project.ToolsDirectory);
+            log.WriteLine(buildDll ? "LibraryFilename      {0}" : "ExecutableFilename   {0}", project.AssemblyName);
+            log.WriteLine("csc.exe Path         {0}", project.CscPath);
+            log.WriteLine("output path          {0}", project.OutputAssemblyPath);
+            log.WriteList(project.Sources, "SOURCES");
+            log.WriteList(project.Packages, "PACKAGES");
+            log.WriteList(project.References, "REFERENCES");
+            log.WriteList(project.Dependencies, "DEPENDENCIES");
+            log.WriteList(project.CscOptions, "CSCOPTIONS");
+            log.WriteLine("-------------------------------------------------");
+        }
+
+        private static void Adjust(ProjectProperties properties, string adjustmentFilePath, ICollection<string> list)
+        { 
+            if (!File.Exists(adjustmentFilePath)) return;
+            foreach (var line in File.ReadAllLines(adjustmentFilePath))
+            {
+                if (string.IsNullOrWhiteSpace(line)) continue;
+                if (line.StartsWith("//")) continue; // commented out line
+                var adjustment = line.Substring(1).Trim();
+                if (line.StartsWith("-"))
+                {
+                    list.Remove(Path.Combine(properties.PackagesDirectory, adjustment));
+                }
+                else
+                {
+                    list.Add(Path.Combine(properties.PackagesDirectory, adjustment));
+                }
+            }
+        }
+
+        private static List<string> ParseProjectFile(ProjectProperties properties, string projectFile,
+            string elementName)
+        {
+            var attributes = GetAttributes(elementName);
+            if (!File.Exists(projectFile) || attributes == null)
```

Comment:
initialize attributeValues before this block. Then just return it if the condition is met.